src/main/java/org/openjdk/jextract/impl/NameMangler.java

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.jextract.impl;
+
+import org.openjdk.jextract.Declaration;
+import org.openjdk.jextract.Type;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import javax.lang.model.SourceVersion;
+
+/*
+ * This visitor handles java safe names for identifiers, type names. This visitor
+ * stores java safe names in maps. Subsequent code generation steps can check for
+ * java safe names via lookup methods.
+ *
+ * NOTE: Unlike other transforming tree visitors, this one is just identify
+ * transformer. This visitor collects name mappings as it visits tree nodes.
+ * Subsequent code generation steps can check the collected names using getters.
+ */
+final class NameMangler implements TreeTransformer, Declaration.Visitor<Void, Declaration> {
```

Comment:
As discussed offline, there is still coupling between this class and OutputFactory. We have tried several approaches none of which "clicked", so we ended up collecting state in the mangler.

A cleaner solution would be to decorate declarations with "name" attributes, but adding an attribute creates a new declaration, which means there will be types in the tree which would need fixing up. Which is also tricky.